--- a/deep_gemm/include/deep_gemm/scheduler/gemm.cuh
+++ b/deep_gemm/include/deep_gemm/scheduler/gemm.cuh
@@ -38,6 +38,7 @@
           uint32_t kNum1DBlocksPerGroup = get_num_1d_blocks_per_group<kGemmType, BLOCK_M, BLOCK_N, kNumSMs, kIsMulticastOnA>()>
 struct Scheduler {
     int current_iter = -1;
+    uint32_t cta_idx;

     // Block configs
     uint32_t num_blocks;
@@ -86,7 +87,10 @@

     // ReSharper disable once CppPossiblyUninitializedMember
     CUTLASS_DEVICE explicit Scheduler(const uint32_t& shape_m, const uint32_t& shape_n,
-                                       const uint32_t& shape_k, int* grouped_layout = nullptr) {
+                                       const uint32_t& shape_k, int* grouped_layout = nullptr,
+                                       const uint32_t logical_cta_idx = UINT32_MAX) {
+        cta_idx = logical_cta_idx == UINT32_MAX ?
+            static_cast<uint32_t>(blockIdx.x) : logical_cta_idx;
         num_m_blocks = math::ceil_div(shape_m, BLOCK_M);
         num_n_blocks = math::ceil_div(shape_n, BLOCK_N);
         current_shape_k = shape_k;
@@ -195,7 +199,7 @@
     }

     CUTLASS_DEVICE bool get_next_block(uint32_t& m_block_idx, uint32_t& n_block_idx) {
-        const auto next_block_idx = (++ current_iter) * kNumSMs + blockIdx.x;
+        const auto next_block_idx = (++ current_iter) * kNumSMs + cta_idx;

         if constexpr (kGemmType == GemmType::MGroupedMasked) {
             while (true) {
